### PR TITLE
Faster connection error detection.

### PR DESCRIPTION
**Pull request by fourtytoo**
Infinite timeout in poll()s.

### CODE REVIEW

File: easywsclient.cpp [hunk 1]
```diff
@@ -188,15 +188,25 @@ class _RealWebSocket : public easywsclient::WebSocket
             }
             return;
         }
-        if (timeout > 0) {
+        if (timeout != 0) {
```
Comment (dhbaird) [inline easywsclient.cpp]: What is the significance of timeout < 0?

File: easywsclient.cpp [hunk 2]
```diff
@@ -188,15 +188,25 @@ class _RealWebSocket : public easywsclient::WebSocket
...
+            fd_set efds;
             timeval tv = { timeout/1000, (timeout%1000) * 1000 };
             FD_ZERO(&rfds);
             FD_ZERO(&wfds);
+            FD_ZERO(&efds);
             FD_SET(sockfd, &rfds);
             if (txbuf.size()) { FD_SET(sockfd, &wfds); }
-            select(sockfd + 1, &rfds, &wfds, NULL, &tv);
+            FD_SET(sockfd, &efds);
+            select(sockfd + 1, &rfds, &wfds, &efds, timeout > 0 ? &tv : 0);
```
Comment (dhbaird) [inline easywsclient.cpp]: Is it okay to remove the timeout > 0 test here, and move it to the if-statement above?

**Comment by dhbaird** [comment]
Hi @fourtytoo, can you review the comments above? This patch looks good if you can address those issues. Thanks!

**Comment by dhbaird** [comment]
Oh, I think I see what you are going for: If timeout < 0, then this code will give better socket error detection that otherwise would be ignored and cause a delay?

**Comment by dhbaird** [comment]
Ookay, I got it- infinite timeout is what you were going for, in addition to the extra errfd checks. This sounds good. Will merge. Would you be able to do a documentation update if you have a chance, e.g., in the README:

    void poll(int timeout = 0); // timeout in milliseconds

to

    void poll(int timeout = 0);
        // Timeout in milliseconds when > 0.
        // Non-blocking when 0.
        // Blocks indefinitely when < 0.

**Comment by dhbaird** [comment]
Yikes!!! Really sorry to flop on this, but I am now unclear about the use case for efds:

    http://stackoverflow.com/questions/1342712/nix-select-and-exceptfds-errorfds-semantics

It says "It is sometimes thought that exceptfds is needed to detect errors, but that is a misconception. Errors will be flagged in readfds ..." . Though I don't really see any harm in using efds, it sounds like it might be extraneous. Could you explain about how your use case is using the efds? Thanks!

**Comment by fourtytoo** [comment]
Actually, I was taken aback when the README actually already documented
that feature:

// Function to perform actual network send()/recv() I/O:// (note: if
all you need is to recv()/dispatch() messages, then timeout can be//
used to block until a message arrives. By default, when timeout is 0,
poll()// will not block at all.)

but I couldn't find a way to specify an endless timeout.

BTW the C++11 version it wasn't touched by my patch.

**Comment by dhbaird** [comment]
Yeah, I agree with you that a block-indefinitely feature would be an improvement, and would like to merge this patch. Would you be able to address the question above about the use case for efds?

**Comment by fourtytoo** [comment]
That's exactly what I learned from the Stevens' book, but it wasn't the
behavior of OSX.
Actually, in our use case, the server was running on a virtual machine;
without select(), the "power-off" of the machine wasn't detected until
minutes later.
Frankly, I don't know what to say.  Just leave out that part if you aren't
comfortable with that.

**Comment by dhbaird** [comment]
Naah, I won't leave it out. But I want to have an idea of why it is in there in case someone else wants to take it out, and I can justify it to them! :)

**Comment by dhbaird** [comment]
Are you sure that with efds, in OS X, it actually does give it the ability to detect that the other machine was powered off?

**Comment by fourtytoo** [comment]
I don't have the figures with me, but I believe tonight I'll be able to
repeat some tests with a sample program.
Please bear with me.

**Comment by dhbaird** [comment]
Awesome; I'm patient. Also, thanks a lot for looking into this. I really appreciate it!

**Comment by fourtytoo** [comment]
It seems that either way (with efds and without) the unplug of the server
running on the virtual machine is not detected.
The test has been done with poll(-1).  With a timeout, the death of the
socket is detected with or without efds.
Ergo, there is no difference, and the Stevens' can be put back on the altar.

So it appears that our tests had been done in different circumstances (with
and without timeout, I suspect).
Sorry about that.

**Comment by dhbaird** [comment]
No worries! I'd still like to merge the indefinite timeout though. If you are still interested, would you be able to re-submit the patch with just that part?

**Comment by fourtytoo** [comment]
You should have received the pull req.

**Comment by dhbaird** [comment]
Merged #43. Thanks!

**Pull request closed without merging**